common/nymsphinx/src/preparer/mod.rs

```diff
@@ -362,108 +282,183 @@ where
     }
 
-    fn deterministic_route_selection(&self) -> bool {
-        self.deterministic_route_selection
-    }
+    fn prepare_chunk_for_sending(
+        &mut self,
+        fragment: Fragment,
+        topology: &NymRouteProvider,
+        ack_key: &AckKey,
+        packet_recipient: &Recipient,
+        packet_type: PacketType,
+    ) -> Result<PreparedFragment, NymTopologyError> {
+        debug!("Preparing chunk for sending");
+        // each plain or repliable packet (i.e. not a reply) attaches an ephemeral public key so that the recipient
+        // could perform diffie-hellman with its own keys followed by a kdf to re-derive
+        // the packet encryption key
 
-    fn rng(&mut self) -> &mut Self::Rng {
-        &mut self.rng
-    }
+        // let fragment_header = fragment.header();
+        let destination = packet_recipient.gateway();
+        monitoring::fragment_sent(&fragment, self.nonce(), destination);
 
-    fn nonce(&self) -> i32 {
-        self.nonce
-    }
+        let non_reply_overhead = encryption::PUBLIC_KEY_SIZE;
+        let expected_plaintext = match packet_type {
+            PacketType::Outfox => {
+                fragment.serialized_size() + OUTFOX_ACK_OVERHEAD + non_reply_overhead
+            }
+            _ => fragment.serialized_size() + ACK_OVERHEAD + non_reply_overhead,
+        };
 
-    fn average_packet_delay(&self) -> Duration {
-        self.average_packet_delay
-    }
+        // the reason we're unwrapping (or rather 'expecting') here rather than handling the error
+        // more gracefully is that this error should never be reached as it implies incorrect chunking
+        let packet_size = PacketSize::get_type_from_plaintext(expected_plaintext, packet_type)
+            .expect("the message has been incorrectly fragmented");
 
-    fn average_ack_delay(&self) -> Duration {
-        self.average_ack_delay
+        let fragment_identifier = fragment.fragment_identifier();
+
+        // create an ack
+        let surb_ack =
+            self.generate_surb_ack(fragment_identifier, topology, ack_key, packet_type)?;
+        let ack_delay = surb_ack.expected_total_delay();
+
+        let packet_payload = match NymPayloadBuilder::new(fragment, surb_ack)
+            .build_regular(self.rng(), packet_recipient.encryption_key())
+        {
+            Ok(payload) => payload,
+            Err(_e) => return Err(NymTopologyError::PayloadBuilder),
+        };
+
+        trace!("Preparing chunk for sending");
+        let route = if self.disable_mix_hops {
+            topology.empty_route_to_egress(destination)?
+        } else {
+            // generate pseudorandom route for the packet
+            topology.random_route_to_egress(self.rng(), destination)?
+        };
+
+        let destination = packet_recipient.as_sphinx_destination();
+
+        // including set of delays
+        let delays =
+            nym_sphinx_routing::generate_hop_delays(self.average_packet_delay(), route.len());
```

Comment:
this is a side note but makes me wonder whether there's a bug in delay generation. it's because we're generating 1 delay here, but, when egress receives the packet and unwraps the header, the delay information wouldn't be there. the whole content would be immediately forwarded to the recipient... 🤔

Reply:
I don't know the answer to this, is this worth opening an issue? Not sure this is directly related to this PR